src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,264 @@
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void BuildRenderTree(RenderTreeBuilder builder)
+        {
+            var itemsAfter = Math.Max(0, _itemCount - _itemsVisible - _itemsBefore);
+
+            builder.OpenElement(0, "div");
+            builder.AddAttribute(1, "style", GetSpacerStyle(_itemsBefore));
+            builder.AddElementReferenceCapture(2, elementReference => _spacerBefore = elementReference);
+            builder.CloseElement();
+
+            builder.OpenRegion(3);
+
+            var itemsRendered = 0;
+
+            // Render as many items as will visibly fit.
+            foreach (var item in _loadedItems.Skip(_itemsBefore).Take(_itemsVisible))
+            {
+                builder.AddContent(0, _itemTemplate, item);
+                itemsRendered++;
+            }
+
+            // Render placholder items to fill the remaining space.
+            for (; itemsRendered < _itemsVisible; itemsRendered++)
+            {
+                builder.AddContent(0, _placeholder, _itemsBefore + itemsRendered);
+            }
+
+            builder.CloseRegion();
+
+            builder.OpenElement(4, "div");
+            builder.AddAttribute(5, "style", GetSpacerStyle(itemsAfter));
+            builder.AddElementReferenceCapture(6, elementReference => _spacerAfter = elementReference);
+
+            builder.CloseElement();
+        }
+
+        private string GetSpacerStyle(int itemsInSpacer)
+            => $"height: {itemsInSpacer * ItemSize}px;";
+
+        void IVirtualizeJsCallbacks.OnBeforeSpacerVisible(float spacerSize, float containerSize)
+        {
+            _itemsBefore = CalcualteItemDistribution(spacerSize, containerSize);
+
+            StateHasChanged();
+        }
+
+        void IVirtualizeJsCallbacks.OnBottomSpacerVisible(float spacerSize, float containerSize)
+        {
+            var itemsAfter = CalcualteItemDistribution(spacerSize, containerSize);
+
+            _itemsBefore = Math.Max(0, _itemCount - itemsAfter - _itemsVisible);
+
+            if (_itemsBefore + _visibleItemCapacity > _loadedItems.Count)
+            {
+                _fetchTask = FetchItems();
+            }
+
+            StateHasChanged();
+        }
+
+        private int CalcualteItemDistribution(float spacerSize, float containerSize)
+        {
+            _visibleItemCapacity = (int)Math.Ceiling(containerSize / ItemSize) + 2;
+            _itemsVisible = Math.Clamp(_itemCount, 0, _visibleItemCapacity);
+
+            return Math.Max(0, (int)Math.Floor(spacerSize / ItemSize) - 1);
+        }
+
+        private async Task FetchItems()
+        {
+            // Cancel the previous fetch, if it exists.
+            _fetchCts?.Cancel();
+
+            // Wait for the task to complete. If it fails due to an exception, the exception will bubble-up
+            // through `OnAfterRenderAsync`.
+            if (!_fetchTask.IsCanceled)
+            {
+                await _fetchTask;
+            }
```

Comment:
I suspect we'll also need to handle the case where `_fetchTask` goes on to get cancelled. With the implementation as-is, the `await _fetchTask` would throw in this case, whereas I think we'd want it to no-op. So something like:

```cs
if (!_fetchTask.IsCompletedSuccessfully)
{
    try
    {
        await _fetchTask;
    }
    catch
    {
        if (!task.IsCanceled)
        {
            throw;
        }
    }
}
```

This is basically the same as the cancellation handling we have in `ComponentBase`.